scipy/stats/_stats_mstats_common.py

```diff
@@ -135,55 +141,53 @@ def theilslopes(y, x=None, alpha=0.95, method='separate'):
     if method not in ['joint', 'separate']:
         raise ValueError("method must be either 'joint' or 'separate'."
                          f"'{method}' is invalid.")
-    # We copy both x and y so we can use _find_repeats.
-    y = np.array(y, dtype=float, copy=True).ravel()
-    if x is None:
-        x = np.arange(len(y), dtype=float)
-    else:
-        x = np.array(x, dtype=float, copy=True).ravel()
-        if len(x) != len(y):
-            raise ValueError("Array shapes are incompatible for broadcasting.")
-    if len(x) < 2:
-        raise ValueError("`x` and `y` must have length at least 2.")
+
+    y, x = xp_promote(y, x, force_floating=True, xp=np)
+    x = np.arange(y.shape[-1], dtype=y.dtype) if x is None else x
+    y, x = np.broadcast_arrays(y, x)
 
     # Compute sorted slopes only when deltax > 0
-    deltax = x[:, np.newaxis] - x
-    deltay = y[:, np.newaxis] - y
-    slopes = deltay[deltax > 0] / deltax[deltax > 0]
-    if not slopes.size:
-        msg = "All `x` coordinates are identical."
-        warnings.warn(msg, RuntimeWarning, stacklevel=2)
-    slopes.sort()
-    medslope = np.median(slopes)
+    deltax = x[..., :, np.newaxis] - x[..., np.newaxis, :]
+    deltay = y[..., :, np.newaxis] - y[..., np.newaxis, :]
+    i = np.triu(np.ones(deltax.shape[-2:], dtype=bool), k=1)
+    # with NumPy:
+    deltax, deltay = deltax[..., i], deltay[..., i]
+    # with array API, mask must be sole index, so we'll need to broadcast it.
+    # indexing will ravel the array, so we'll need to reshape it after
+    deltax[deltax == 0] = np.nan
+    slopes = deltay / deltax
+    slopes = np.sort(slopes, axis=-1)
+    medslope = np.nanmedian(slopes, axis=-1)
     if method == 'joint':
-        medinter = np.median(y - medslope * x)
+        medinter = np.nanmedian(y - medslope * x, axis=-1)
     else:
-        medinter = np.median(y) - medslope * np.median(x)
+        medinter = np.nanmedian(y, axis=-1) - medslope * np.nanmedian(x, axis=-1)
     # Now compute confidence intervals
     if alpha > 0.5:
         alpha = 1. - alpha
 
     z = distributions.norm.ppf(alpha / 2.)
     # This implements (2.6) from Sen (1968)
-    _, nxreps = _find_repeats(x)
-    _, nyreps = _find_repeats(y)
-    nt = len(slopes)       # N in Sen (1968)
-    ny = len(y)            # n in Sen (1968)
+    _, nxreps = _stats_py._rankdata(x, method='average', return_ties=True)
+    _, nyreps = _stats_py._rankdata(y, method='average', return_ties=True)
```

Comment:
The second output of `_find_repeats` is like the counts of `unique_counts`. The second output of `return_ties` is similar, but it is has zeros to keep the array rectangular. Because the tie correction formula has terms like `nxreps * (nxreps-1)`, we get the same result.